gettsim/taxes/zve.py

```diff
@@ -313,57 +318,145 @@ def calc_gross_e7(tax_unit, params):
+            sonstigevors2010 + 12 * tax_unit["avbeit"], params["vors_sonst_max"]
         ),
     )
+    return altersvors2010.astype(int) + sonstigevors2010.astype(int)
 
-    vorsorge2010 = altersvors2010 + sonstigevors2010
 
-    return vorsorge2010.astype(int)
+def vorsorge2005(tax_unit, params, soz_vers_beitr_params):
+    """ Vorsorgeaufwendungen 2005 to 2010
+    Pension contributions are accounted for up to €20k.
+    From this, a certain share can actually be deducted,
+    starting with 60% in 2005.
+    Other deductions are simply added up, up to a ceiling of 1500 p.a. for standard employees.
+    """
+
+    altersvors2005 = calc_altersvors_aufwend(tax_unit, params)
+
+    sonstigevors2005 = ~tax_unit["child"] * np.minimum(
+        params["vors_sonst_max"],
+        12 * (tax_unit["gkvbeit"] + tax_unit["avbeit"] + tax_unit["pvbeit"]),
+    ).astype(int)
 
+    return (altersvors2005 + sonstigevors2005).astype(int)
+
+
+def vorsorge2004(tax_unit, params, soz_vers_beitr_params):
+    """ Vorsorgeaufwendungen up until 2004.
+        - only pension and health contributions.
+    """
 
-def vorsorge_dummy(tax_unit, params, soz_vers_beitr_params):
-    return 0
+    # Distinguish between married and singles
+    # Single Taxpayer
+    if not tax_unit["zveranl"].max():
+        # Amount 1: Basic deduction, based on earnings. Usually zero.
+        item_1 = np.maximum(
+            params["vorwegab"] - params["kuerzquo"] * 12 * tax_unit["m_wage"], 0
+        )
+        # calcuate the remaining amount.
+        vorsorg_rest = np.maximum(
+            12 * (tax_unit["rvbeit"] + tax_unit["gkvbeit"]) - item_1, 0
+        )
+        # Deduct a 'Grundhöchstbetrag' (1334€ in 2004),
+        # or the actual expenses if lower (which is unlikely)
+        item_2 = np.minimum(params["grundbet"], vorsorg_rest)
+        # From what is left from vorsorg_rest, you may deduct 50%.
+        # (up until 50% of 'Grundhöchstbetrag')
+        item_3 = np.minimum(0.5 * (vorsorg_rest - item_2), 0.5 * params["grundbet"])
+    # For the married couple, the same stuff, but with tu totals.
+    if tax_unit["zveranl"].max():
+        for var in ["m_wage", "rvbeit", "gkvbeit"]:
+            tax_unit[f"{var}_tu"] = tax_unit.loc[~tax_unit["child"], "m_wage"].sum()
+
+        item_1 = 0.5 * np.maximum(
+            2 * params["vorwegab"] - params["kuerzquo"] * 12 * tax_unit["m_wage_tu"], 0
+        )
+        vorsorg_rest = 0.5 * np.maximum(
+            12 * (tax_unit["rvbeit_tu"] + tax_unit["gkvbeit_tu"]) - item_1, 0
+        )
+        item_2 = 0.5 * np.minimum(params["grundbet"], vorsorg_rest)
+        item_3 = 0.5 * np.minimum((vorsorg_rest - item_2), 2 * params["grundbet"])
+
+    # Finally, add up all three amounts and assign in to the adults.
+    vorsorge2004 = ~tax_unit["child"] * (item_1 + item_2 + item_3).astype(int)
+
+    return vorsorge2004
+
+
+def vorsorge04_05(tax_unit, params, soz_vers_beitr_params):
```

Comment:
The names are trying to give too much information. Without looking at the functions, it is unclear what this means.

I would suggest the following names:
```
vorsorge_pre_2005 (=current 2004)
vorsorge_since_2005 (=current 04_05, 04_10)
_vorsorge_since_2005 (current 2005 / 2010)
```

Btw, does the max(04, x)-thing apply also to insurance contracts that were made after 2005? Or only to old ones?

Reply:
OK with refactoring. It does not matter when insurance contracts started. Only the year in which contributions are paid matter.